shared/dataflow/codeql/dataflow/VariableCapture.qll

```diff
@@ -0,0 +1,930 @@
+ * for supporting flow through captured variables.
+ */
+
+private import codeql.util.Boolean
+private import codeql.util.Unit
+private import codeql.ssa.Ssa as Ssa
+
+signature module InputSig {
+  class Location {
+    predicate hasLocationInfo(
+      string filepath, int startline, int startcolumn, int endline, int endcolumn
+    );
+  }
+
+  /**
+   * A basic block, that is, a maximal straight-line sequence of control flow nodes
+   * without branches or joins.
+   */
+  class BasicBlock {
+    /** Gets a textual representation of this basic block. */
+    string toString();
+
+    /** Gets the enclosing callable. */
+    Callable getEnclosingCallable();
+
+    /** Gets the location of this basic block. */
+    Location getLocation();
+  }
+
+  /**
+   * Gets the basic block that immediately dominates basic block `bb`, if any.
+   *
+   * That is, all paths reaching `bb` from some entry point basic block must go
+   * through the result.
+   *
+   * Example:
+   *
+   * ```csharp
+   * int M(string s) {
+   *   if (s == null)
+   *     throw new ArgumentNullException(nameof(s));
+   *   return s.Length;
+   * }
+   * ```
+   *
+   * The basic block starting on line 2 is an immediate dominator of
+   * the basic block on line 4 (all paths from the entry point of `M`
+   * to `return s.Length;` must go through the null check.
+   */
+  BasicBlock getImmediateBasicBlockDominator(BasicBlock bb);
+
+  /** Gets an immediate successor of basic block `bb`, if any. */
+  BasicBlock getABasicBlockSuccessor(BasicBlock bb);
+
+  /** Holds if `bb` is a control-flow entry point. */
+  default predicate entryBlock(BasicBlock bb) { not exists(getImmediateBasicBlockDominator(bb)) }
+
+  /** Holds if `bb` is a control-flow exit point. */
+  default predicate exitBlock(BasicBlock bb) { not exists(getABasicBlockSuccessor(bb)) }
+
+  /** A variable that is captured in a closure. */
+  class CapturedVariable {
+    /** Gets a textual representation of this variable. */
+    string toString();
+
+    /** Gets the callable that defines this variable. */
+    Callable getCallable();
+
+    /** Gets the location of this variable. */
+    Location getLocation();
+  }
+
+  /** A parameter that is captured in a closure. */
+  class CapturedParameter extends CapturedVariable;
+
+  /**
+   * An expression with a value. That is, we expect these expressions to be
+   * represented in the data flow graph.
+   */
+  class Expr {
+    /** Gets a textual representation of this expression. */
+    string toString();
+
+    /** Gets the location of this expression. */
+    Location getLocation();
+
+    /** Holds if the `i`th node of basic block `bb` evaluates this expression. */
+    predicate hasCfgNode(BasicBlock bb, int i);
+  }
+
+  /** A write to a captured variable. */
+  class VariableWrite {
+    /** Gets the variable that is the target of this write. */
+    CapturedVariable getVariable();
+
+    /** Gets the expression that is the source of this write. */
```

Comment:
Hmm, what's a good example where there's no source? Currently I have put in a consistency check that there's supposed to be exactly one source of a write, but there may very well be cases I haven't thought of.

Reply:
For example `out/ref` arguments in C#.

Reply:
But supposedly those writes need incoming flow from the output of those arguments? And that incoming flow will have associated some `DataFlow::Node`. This is actually likely a good example where it isn't sufficient to simply give the C# class `Expr` as the input class `Expr` here - instead you'll possibly need to provide a newtype in order to include something for `out/ref` arguments, so that you'll actually be able to get that flow as well.
Another example is the enhanced for statements of Java, these also don't have Java `Expr` sources for the writes, so that's why the capture of loop variables from enhanced for statements currently doesn't work (documented with a test, and I've made and issue to track the followup, and the corresponding consistency check also highlights this problem). In order to fix this, I'll need to replace the `Expr` input to be a disjoint union of Java `Expr` and something for enhanced for statements. I'll likely also need to add an additional intermediate `DataFlow::Node`.